hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -0,0 +1,399 @@
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import static com.hazelcast.util.ExceptionUtil.rethrow;
+import static com.hazelcast.util.Preconditions.isNotNull;
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+import static java.util.concurrent.locks.LockSupport.park;
+import static java.util.concurrent.locks.LockSupport.parkNanos;
+import static java.util.concurrent.locks.LockSupport.unpark;
+
+/**
+ * todo:
+ * - thread value protection
+ *
+ * @param <V>
+ */
+public abstract class AbstractInvocationFuture<V> implements InternalCompletableFuture<V> {
+
+    static final Object VOID = "VOID";
+
+    private static final AtomicReferenceFieldUpdater<AbstractInvocationFuture, Object> STATE =
+            newUpdater(AbstractInvocationFuture.class, Object.class, "state");
+
+    protected final Executor defaultExecutor;
+    protected final ILogger logger;
+
+    /**
+     * This field contain the state of the future. Either the future is not complete and the state is:
+     * <ol>
+     * <li>VOID: no response is available.</li>
+     * <li>Thread instance: no response is available and a thread has blocked on completion (e.g. future.get)</li>
+     * <li>{@link ExecutionCallback} instance: no response is available and 1 {@link #andThen(ExecutionCallback)}
+     * was done using the default executor</li>
+     * <li>{@link WaitNode} instance: in case of multiple andThen registrations or future.gets or andThen with custom Executor
+     * </li>
+     * </ol>
+     *
+     * If the state is anything else, it is completed.
+     *
+     * The reason why a single future.get or registered ExecutionCallback doesn't create a WaitNode is that we don't want to cause
+     * additional litter since most of our API calls are a get or a single ExecutionCallback.
+     *
+     * The state field is replaced using a cas, so registration or setting a response is an atomic operation and therefor not
+     * prone to data-races. There is no need to use synchronized blocks.
+     */
+    private volatile Object state = VOID;
+
+    public AbstractInvocationFuture(Executor defaultExecutor, ILogger logger) {
+        this.defaultExecutor = defaultExecutor;
+        this.logger = logger;
+    }
+
+    boolean compareAndSetState(Object oldState, Object newState) {
+        return STATE.compareAndSet(this, oldState, newState);
+    }
+
+    protected final Object getState() {
+        return state;
+    }
+
+    @Override
+    public final boolean isDone() {
+        return isDone(state);
+    }
+
+    private boolean isDone(final Object state) {
+        if (state == null) {
+            return true;
+        }
+
+        return !(state == VOID
+                || state.getClass() == WaitNode.class
+                || state instanceof Thread
+                || state instanceof ExecutionCallback);
+    }
+
+    protected void onInterruptDetected() {
```

Comment:
Because they behave differently. Because a shared implementation is used, this inconsistency becomes visible. I would like to get this fixed; but not as part of this PR. There are also other inconsistencies e.g. andThen exception handling. On member, the ExecutionExeception envelope is removed, on the client not.